### PR TITLE
A MongoDB Component Downloader

### PR DESCRIPTION
**Pull request by vector-of-bool**
This adds a self-contained Python program that downloads MongoDB components based on the contents of `full.json`. The downloads and the download manifest are cached locally, and items can be selectively extracted from the archive files. Availability of components is query-able using the `--list` flag.

The `--help` output, for reference:

```
usage: mongodl.py [-h] [--cache-dir CACHE_DIR] [--list] [--target TARGET] [--arch ARCH]
                  [--edition EDITION] [--out OUT] [--version VERSION]
                  [--component COMPONENT] [--only ONLY] [--strip-path-components N]
                  [--test] [--empty-is-error]

optional arguments:
  -h, --help            show this help message and exit
  --cache-dir CACHE_DIR
                        Directory where download caches and metadata will be stored

List arguments:
  --list                List available components, targets, editions, and architectures.
                        Download arguments will act as filters.

Download arguments:
  Select what to download and extract. Non-required arguments will be inferred based on
  the host system.

  --target TARGET, -T TARGET
                        The target platform for which to download. Use "--list" to list
                        available targets.
  --arch ARCH, -A ARCH  The architecture for which to download
  --edition EDITION, -E EDITION
                        The edition of the product to download (Default is "enterprise").
                        Use "--list" to list available editions.
  --out OUT, -o OUT     The directory in which to download components. (Required)
  --version VERSION, -V VERSION
                        The product version to download (Required). Use "--list" to list
                        available versions.
  --component COMPONENT, -C COMPONENT
                        The component to download (Required). Use "--list" to list
                        available components.
  --only ONLY           Restrict extraction to items that match the given globbing
                        expression. The full archive member path is matched, so a pattern
                        like "*.exe" will only match "*.exe" at the top level of the
                        archive. To match recursively, use the "**" pattern to match any
                        number of intermediate directories.
  --strip-path-components N, -p N
                        Strip the given number of path components from archive members
                        before extracting into the destination. The relative path of the
                        archive member will be used to form the destination path. For
                        example, a member named [bin/mongod.exe] will be extracted to
                        [<out>/bin/mongod.exe]. Using --strip-components=1 will remove the
                        first path component, extracting such an item to
                        [<out>/mongod.exe]. If the path has fewer than N components, that
                        archive member will be ignored.
  --test                Do not extract or place any files/directories. Only print what
                        will be extracted without placing any files.
  --empty-is-error      If all files are excluded by other filters, treat that situation
                        as an error and exit non-zero.
```

### CODE REVIEW

**Comment by kevinAlbs** [comment]
> @kevinAlbs is there anyone else that might want to review this?

Requested @ShaneHarvey since he is [a co-owner of Evergreen Tools](https://wiki.corp.mongodb.com/pages/viewpage.action?pageId=116568903).

**Review by ShaneHarvey — COMMENTED** [review]
I believe [curator](https://github.com/mongodb/curator) already has this functionality. Have you tried using curator instead?

**Comment by vector-of-bool** [comment]
@ShaneHarvey This was actually spawned from DRIVERS-2241, which included a sub-task to write a script to download the `csfle` DSO automatically. It involved pulling `full.json`, downloading the archive, then expanding it and picking out files, and so I just generalized the task to download any arbitrary components.

While not directed specifically at DRIVERS-1827, it may be useful for that since it will automatically see new entries in `full.json`. Like `download-mongodb.sh`, the intent is that drivers can copy this script into their respective repositories for use in automation and tooling. This program is very much designed to be easily tool-driven, hence the many path-manipulation options.

**Comment by ShaneHarvey** [comment]
Ah okay so this is related to csfle, but I'm still a little confused. [curator](https://github.com/mongodb/curator) is already able to download mongodb binaries and parse full.json. Shouldn't we add csfle download support to curator itself rather than build a new tool?

**Review by ShaneHarvey — COMMENTED** [review]
Note please don't hold up this review on my account. I briefly glanced at the code and it LGTM. I'll defer to Kevin's review. Nice work!

File: .evergreen/mongodl.py
Comment (ShaneHarvey) [reply .evergreen/mongodl.py:1]: Might be better to make it "python3" instead of "python". And maybe with a fallback to one of the python or server toolchains:
```
C:/python/Python39/python.exe
/opt/mongodbtoolchain/v3/bin/python3
/opt/python/3.9/bin/python3
```

Comment (vector-of-bool) [reply .evergreen/mongodl.py:1]: @kevinAlbs I added a `--verison=latest` feature, although that may be different than what you're looking for since that will auto-update the moment a new `full.json` is published.

If the intent is to make it as easy as possible to update the pulled `csfle` version, would it not be at least as easy to change the `--version=` option used in a build script as it would be to update their copy of the drivers-evergreen-tools repo?

Comment (ShaneHarvey) [reply .evergreen/mongodl.py:1]: Driver tests automatically install drivers-evergreen-tools@master so the advantage is that we can update drivers-evergreen-tools once and the change will be automatically applied to all driver tests.

Comment (kevinAlbs) [reply .evergreen/mongodl.py:1]: I like the `--version=latest` option. Language drivers can use that by default. In the rare scenario that a driver needs to pin to an older release, they could still do so in their respective repos.

File: .evergreen/mongodl.py
Comment (kevinAlbs) [inline .evergreen/mongodl.py:217]: IIUC this currently considers `rc < alpha < beta` for getting the latest release. I think that may need to be flipped. `rc` is the most stable.

File: .evergreen/mongodl.py
Comment (kevinAlbs) [inline .evergreen/mongodl.py:203]:
```suggestion
#: Regular expression that matches the version numbers from 'full.json'
```

**Review by kevinAlbs — APPROVED** [review]
LGTM with a possible change to the tag ordering. Nicely done!

**Pull request merged**
